### PR TITLE
[animation] High-level animations return cancelable interface

### PR DESCRIPTION
**Pull request by kiryldz**
PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [x] Include before/after visuals or gifs if this PR includes visual changes.
 - [x] Write tests for all new functionality. If tests were not written, please explain why.
 - [x] Add example if relevant.
 - [x] Document any changes to public APIs.
 - [x] Apply changelog label ('breaking change', 'bug :beetle:', 'build', 'docs', 'feature :green_apple:', 'performance :zap:', 'testing :100:') or use the label 'skip changelog'
 - [x] Add an entry inside this element for inclusion in the `mapbox-maps-android` changelog: `<changelog>[animation] High-level animations return cancelable interface</changelog>`.

### Summary of changes

High-level animation functions like `flyTo`, `easeTo` etc now return `Cancelable` interface allowing to cancel that high-level animation if needed.

### CODE REVIEW

File: app/src/main/res/layout/activity_camera_predefined_animators.xml [hunk 1]
```diff
@@ -11,18 +10,31 @@
...
 
+    <Button
+        android:id="@+id/buttonCancel"
+        android:visibility="invisible"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginEnd="50dp"
+        android:layout_marginBottom="20dp"
+        android:padding="10dp"
+        android:text="Cancel animation"
```
Comment (Chaoba) [inline app/src/main/res/layout/activity_camera_predefined_animators.xml:35]: Can you define the string in string.xml?

Comment (kiryldz) [reply app/src/main/res/layout/activity_camera_predefined_animators.xml:35]: Done already, lint fails without it 😄

**Review by Chaoba — APPROVED** [review]
LGTM

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt [hunk 2]
```diff
@@ -557,12 +557,14 @@ internal class CameraAnimationsPluginImpl : CameraAnimationsPlugin {
...
+   *
+   * @return [Cancelable] animator set object.
    */
   override fun easeTo(
     cameraOptions: CameraOptions,
     animationOptions: MapAnimationOptions?
-  ) {
-    startHighLevelAnimation(
+  ): Cancelable {
+    return startHighLevelAnimation(
```
Comment (pengdev) [inline plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:567]:
```suggestion
  ): Cancelable = startHighLevelAnimation(
```

File: app/src/main/java/com/mapbox/maps/testapp/examples/CameraPredefinedAnimatorsActivity.kt [hunk 3]
```diff
@@ -132,48 +139,83 @@ class CameraPredefinedAnimatorsActivity : AppCompatActivity() {
...
-      R.id.menu_action_ease_to -> mapboxMap.easeTo(
-        EASE_TO_TARGET_CAMERA_POSITION,
-        mapAnimationOptions {
-          duration(2000)
+      R.id.menu_action_ease_to ->
+        mapboxMap.easeTo(
+          EASE_TO_TARGET_CAMERA_POSITION,
+          prepareAnimationOptions(2000)
+        )?.let {
+          runningCancelableAnimator = it
```
Comment (pengdev) [inline app/src/main/java/com/mapbox/maps/testapp/examples/CameraPredefinedAnimatorsActivity.kt:180]:
```suggestion
        runningCancelableAnimator = mapboxMap.easeTo(
          EASE_TO_TARGET_CAMERA_POSITION,
          prepareAnimationOptions(2000)
        )
```

File: app/src/main/java/com/mapbox/maps/testapp/examples/CameraPredefinedAnimatorsActivity.kt [hunk 4]
```diff
@@ -132,48 +139,83 @@ class CameraPredefinedAnimatorsActivity : AppCompatActivity() {
...
-      )
-      R.id.menu_action_fly_to -> mapboxMap.flyTo(
-        EASE_TO_TARGET_CAMERA_POSITION,
-        mapAnimationOptions {
-          duration(2000)
+      R.id.menu_action_fly_to ->
+        mapboxMap.flyTo(
+          EASE_TO_TARGET_CAMERA_POSITION,
+          prepareAnimationOptions(2000)
+        )?.let {
```
Comment (pengdev) [inline app/src/main/java/com/mapbox/maps/testapp/examples/CameraPredefinedAnimatorsActivity.kt:186]: why the mapboxMap.flTo() returns a nullable?

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt [hunk 5]
```diff
@@ -803,35 +818,41 @@ fun MapPluginProviderDelegate.getCameraAnimationsPlugin(): CameraAnimationsPlugi
...
  * @param cameraOptions The camera options to ease to
  * @param animationOptions Transition options (animation duration, listeners etc)
+ *
+ * @return [Cancelable] animator set object or null if associated map object was garbage collected.
  */
 fun MapPluginExtensionsDelegate.easeTo(
   cameraOptions: CameraOptions,
   animationOptions: MapAnimationOptions? = null
-) = cameraAnimationsPlugin { easeTo(cameraOptions, animationOptions) }
+) = cameraAnimationsPlugin { easeTo(cameraOptions, animationOptions) } as Cancelable?
```
Comment (pengdev) [inline plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:827]: Why would it return a nullable?

Comment (kiryldz) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:827]: Good catch, should not be nullable 👍

File: app/src/main/java/com/mapbox/maps/testapp/examples/CameraPredefinedAnimatorsActivity.kt (hunk 4 above)
Comment (kiryldz) [reply app/src/main/java/com/mapbox/maps/testapp/examples/CameraPredefinedAnimatorsActivity.kt:186]: Good catch, should not be nullable.

File: plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt [hunk 6]
```diff
@@ -866,9 +879,11 @@ fun MapPluginExtensionsDelegate.moveBy(
...
+ *
+ * @return [Cancelable] animator set object or null if associated map object was garbage collected.
  */
 fun MapPluginExtensionsDelegate.rotateBy(
   first: ScreenCoordinate,
   second: ScreenCoordinate,
   animationOptions: MapAnimationOptions? = null
-) = cameraAnimationsPlugin { rotateBy(first, second, animationOptions) }
\ No newline at end of file
+) = cameraAnimationsPlugin { rotateBy(first, second, animationOptions) } as Cancelable
```
Comment (pengdev) [inline plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:889]: do we need to do a cast as Cancelable here?

Comment (kiryldz) [reply plugin-animation/src/main/java/com/mapbox/maps/plugin/animation/CameraAnimationsPluginImpl.kt:889]: Yes, we do, we use abstraction with `Any?` under the hood.

File: app/src/main/java/com/mapbox/maps/testapp/examples/CameraPredefinedAnimatorsActivity.kt [hunk 7]
```diff
@@ -132,48 +139,75 @@ class CameraPredefinedAnimatorsActivity : AppCompatActivity() {
...
-          duration(2000)
-        }
-      )
-      R.id.menu_action_scale_by -> mapboxMap.scaleBy(
-        15.0,
-        ScreenCoordinate(10.0, 10.0),
-        mapAnimationOptions {
-          duration(2000)
+  private fun prepareAnimationOptions(duration: Long): MapAnimationOptions {
+    return mapAnimationOptions {
```
Comment (pengdev) [inline app/src/main/java/com/mapbox/maps/testapp/examples/CameraPredefinedAnimatorsActivity.kt:143]:
```suggestion
  private fun prepareAnimationOptions(duration: Long) = mapAnimationOptions {
```

**Review by pengdev — APPROVED** [review]

**Pull request merged**